### PR TITLE
Add thanos ruler deployment

### PR DESCRIPTION
**Pull request by pgier**
This is still a WIP but just wanted to create this PR so others can provide feedback.
It looks like we'll need one more release of prometheus-operator before this will work fully.

### CODE REVIEW

**Comment by pgier** [comment]
Unfortuately, some tests are failing due to https://github.com/coreos/prometheus-operator/pull/3045

File: assets/thanos-ruler/stateful-set.yaml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      containers:
+      - args:
+        - rule
+        - --grpc-address=0.0.0.0:10901
+        - --http-address=0.0.0.0:10902
+        - --objstore.config=$(OBJSTORE_CONFIG)
+        - --data-dir=/var/thanos/ruler
+        - --label=ruler_replica="$(NAME)"
+        - --alert.label-drop="ruler_replica"
+        - --query=dnssrv+_http._tcp.thanos-querier.openshift-monitoring.svc.cluster.local
```
Comment (s-urbaniak) [inline assets/thanos-ruler/stateful-set.yaml]: how about the alertmanager configuration?

File: pkg/tasks/thanos_ruler.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+func NewThanosRulerTask(client *client.Client, factory *manifests.Factory) *ThanosRulerTask {
+	return &ThanosRulerTask{
+		client:  client,
+		factory: factory,
+	}
+}
+
+func (t *ThanosRulerTask) Run() error {
+	svc, err := t.factory.ThanosRulerService()
```
Comment (s-urbaniak) [inline pkg/tasks/thanos_ruler.go]: as thanos ruler is part of the user workload monitoring stack, it should only be deployed if that is active.

**Comment by s-urbaniak** [comment]
@pgier looking good so far! So far from a global point I see the following things yet missing:
- Alertmanager integration into ThanosRuler
- Deployment of ThanosRuler if and only if user workload monitoring is enabled
- We need to point ThanosQuerier to ThanosRuler (if it is enabled) such that recording rules can be queried.

**Comment by paulfantom** [comment]
We also need proxy (trusted-ca-bundle) support for this

File: pkg/tasks/thanos_ruler.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		}
+
+		trustedCA, err = t.client.CreateIfNotExistConfigMap(trustedCA)
+		if err != nil {
+			return errors.Wrap(err, "creating Thanos Ruler trusted CA bundle ConfigMap failed")
+		}
+
+		// In the case when there is no data but the ConfigMap is there, we just continue.
+		// We will catch this on the next loop.
+		trustedCA = t.factory.HashTrustedCA(trustedCA, "thanos-ruler")
```
Comment (s-urbaniak) [inline pkg/tasks/thanos_ruler.go]: cc @paulfantom i guess this covers the trusted ca bundle stuff

File: pkg/tasks/thanos_ruler.go (hunk 2 above)
Comment (pgier) [reply pkg/tasks/thanos_ruler.go]: Moved everything over to user workload monitoring

File: pkg/tasks/thanos_ruler.go (hunk 3 above)
Comment (pgier) [reply pkg/tasks/thanos_ruler.go]: I believe all the ca bundle stuff is there and working, but probably still needs some tweaks.

**Comment by lilic** [comment]
Hmm we git some non 2xx in our prometheus tests cc @pgier

**Pull request merged**